Yank/repex.py

```diff
@@ -401,39 +401,37 @@ def write_thermodynamic_states(self, thermodynamic_states, unsampled_states):
         """
         # Store all thermodynamic states as serialized dictionaries.
         compatible_states_hashes = dict()
+        stored_states = dict()
         for state_type, states in [('thermodynamic_states', thermodynamic_states),
                                    ('unsampled_states', unsampled_states)]:
             for state_id, thermodynamic_state in enumerate(states):
-                serialized_state = mmtools.utils.serialize(thermodynamic_state)
-
-                # Find the serialized standard system.
-                serialized_thermodynamic_state = serialized_state
-                while 'thermodynamic_state' in serialized_thermodynamic_state:
-                    # The while loop is necessary for nested CompoundThermodynamicStates.
+                # Check if any compatible state has been found
+                found_compatible_state = False
+                for compare_state in stored_states:
+                    if compare_state.is_state_compatible(thermodynamic_state):
+                        serialized_thermodynamic_state = mmtools.utils.serialize(thermodynamic_state, skip_system=True)
+                        standard_system_hash = stored_states[compare_state]
+                        reference_state_name, len_serialization = compatible_states_hashes[standard_system_hash]
+                        serialized_thermodynamic_state['_Reporter__compatible_state'] = reference_state_name
+                        found_compatible_state = True
+                        break
+                # If no compatible state is found, do full serialization
+                if not found_compatible_state:
+                    serialized_state = mmtools.utils.serialize(thermodynamic_state)
                     serialized_thermodynamic_state = serialized_state['thermodynamic_state']
```

Comment:
I think this will raise `KeyError` for normal `ThermodynamicState`s (only `CompoundThermodynamicState`'s serialization has the `thermodynamic_state` key.